apps/taskmasters/flowlord/taskmaster.go

```diff
@@ -189,55 +189,58 @@ func (tm *taskMaster) schedule() (err error) {
 // Process the given task
 // 1. check if the task needs to be retried
 // 2. start any downstream tasks
-// Send retry failed tasks to tm.failedTopic
+// Send retry failed tasks to tm.failedTopic (only if the phase exists in the workflow)
 func (tm *taskMaster) Process(t *task.Task) error {
 	meta, _ := url.ParseQuery(t.Meta)
 	// attempt to return
 	if t.Result == task.ErrResult {
-		w := tm.Get(*t)
+		p := tm.Get(*t)
 		r := meta.Get("retry")
 		i, _ := strconv.Atoi(r)
-		if w.Retry > i {
-			t = task.NewWithID(t.Type, t.Info, t.ID)
-			i++
-			meta.Set("retry", strconv.Itoa(i))
-			t.Meta = meta.Encode()
-			if err := tm.producer.Send(t.Type, t.JSONBytes()); err != nil {
-				return err
-			}
-		} else if tm.failedTopic != "-" {
-			// send to the retry failed topic if retries > w.Retry
-			meta.Set("retry", "failed")
-			t.Meta = meta.Encode()
-			tm.producer.Send(tm.failedTopic, t.JSONBytes())
-			if tm.slack != nil {
-				b, _ := json.MarshalIndent(t, "", "  ")
-				tm.slack.Notify(string(b), slack.Critical)
+		if p.Task != "" { // the task should have a workflow phase
+			if p.Retry > i {
+				t = task.NewWithID(t.Type, t.Info, t.ID)
+				i++
+				meta.Set("retry", strconv.Itoa(i))
+				t.Meta = meta.Encode()
+				if err := tm.producer.Send(t.Type, t.JSONBytes()); err != nil {
+					return err
+				}
+			} else if tm.failedTopic != "-" {
+				// send to the retry failed topic if retries > p.Retry
+				meta.Set("retry", "failed")
+				t.Meta = meta.Encode()
+				tm.producer.Send(tm.failedTopic, t.JSONBytes())
+				if tm.slack != nil {
+					b, _ := json.MarshalIndent(t, "", "  ")
+					tm.slack.Notify(string(b), slack.Critical)
+				}
```

Comment:
Is there a case where we would want a slack notification, but not to send a message to the failed topic? I think it might make more sense to flatten this logic as well. 

``` go 
return nil 
} // else if tm.failedTopic != "-" {
meta.Set("retry", "failed")
t.Meta = meta.Encode()
if tm.failedTopic != "-" {
		tm.producer.Send(tm.failedTopic, t.JSONBytes())
}
if tm.slack != nil {
	tm.slack.Notify(string(b), slack.Critical)
        b, _ := json.MarshalIndent(t, "", "  ")
	tm.slack.Notify(string(b), slack.Critical)
}
```

Reply:
This is hard to pin down in a real world case.  Flowlord would have to keep track of some kind of state for a specific workflow to know if a message should be sent to slack or not.